### ISSUE TITLE
Default parameters should subtract `undefined` from internal function signature

### ISSUE THREAD

**Issue by blakeembrey** [Microsoft/TypeScript#12726]
**TypeScript Version:**  2.1.4

**Code**

```ts
function f (offset: number) {}

let x = (offset = 0) => {
  f(offset)
}

x = function (offset = 0) {
  f(offset)
}
```

```json
{
  "compilerOptions": {
    "target": "es5",
    "outDir": "dist",
    "module": "commonjs",
    "noImplicitAny": true,
    "strictNullChecks": true
  }
}
```

**Expected behavior:** No error.

**Actual behavior:**

```
index.ts(8,5): error TS2345: Argument of type 'number | undefined' is not assignable to parameter of type 'number'.
  Type 'undefined' is not assignable to type 'number'.
```

I believe this has actually been an issue in previous releases, but 2.1 has made it apparent in other argument positions. I just tested 2.0 and can reproduce the above sample, but it doesn't fail on 2.0 with `(x: string, y = 0)` whereas 2.1 does.

**Comment by mhegazy** [comment]
Looks like the flip side of https://github.com/Microsoft/TypeScript/issues/11900
